### PR TITLE
Fix tuple element tys returned from `sized_conditions()`

### LINKED ISSUE #121443: ICE: `assertion failed: layout.is_sized()`
auto-reduced (treereduce-rust):
````rust
type Fn = FnOnce() -> u8;

const TEST: Fn = my_fn;
const TEST2: (Fn, u8) = (TEST, 0);


````

<details><summary><strong>original code</strong></summary>
<p>

original:
````rust
//@ run-pass

// The actual regression test from #63479. (Including this because my
// first draft at fn-ptr-is-structurally-matchable.rs failed to actually
// cover the case this hit; I've since expanded it accordingly, but the
// experience left me wary of leaving this regression test out.)

#![warn(pointer_structural_match)]

#[derive(Eq)]
struct A {
  a: i64
}

impl PartialEq for A {
    #[inline]
    fn eq(&self, other: &Self) -> bool {
        self.a.eq(&other.a)
    }
}

type Fn = FnOnce() -> u8;

fn my_fn(_args: &[A]) {
  println!("hello world");
}

const TEST: Fn = my_fn;
const TEST2: (Fn, u8) = (TEST, 0);

struct B(Fn);

fn main() {
  let s = B(my_fn);
  match s {
    B(TEST) => println!("matched"),
    //~^ WARN behave unpredictably
    //~| WARN this was previously accepted by the compiler but is being phased out
    _ => panic!("didn't match")
  };
  match (s.0, 0) {
    TEST2 => println!("matched"),
    //~^ WARN behave unpredictably
    //~| WARN this was previously accepted by the compiler but is being phased out
    _ => panic!("didn't match")
  }
}

````
</p>
</details>

Version information
````
rustc 1.78.0-nightly (f8131a48a 2024-02-21)
binary: rustc
commit-hash: f8131a48a46ac3bc8a3d0fe0477055b132cffdc3
commit-date: 2024-02-21
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0425]: cannot find value `my_fn` in this scope
 --> /tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs:3:18
  |
3 | const TEST: Fn = my_fn;
  |                  ^^^^^ not found in this scope

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs:4:35
  |
4 | const TEST2: (Fn, u8) = (TEST, 0);
  |                                   ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs`

warning: trait objects without an explicit `dyn` are deprecated
 --> /tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs:1:11
  |
1 | type Fn = FnOnce() -> u8;
  |           ^^^^^^^^^^^^^^
  |
  = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
  = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
  = note: `#[warn(bare_trait_objects)]` on by default
help: if this is an object-safe trait, use `dyn`
  |
1 | type Fn = dyn FnOnce() -> u8;
  |           +++

error[E0277]: the size for values of type `(dyn FnOnce() -> u8 + 'static)` cannot be known at compilation time
 --> /tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs:3:13
  |
3 | const TEST: Fn = my_fn;
  |             ^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `(dyn FnOnce() -> u8 + 'static)`

error[E0277]: the size for values of type `(dyn FnOnce() -> u8 + 'static)` cannot be known at compilation time
 --> /tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs:3:18
  |
3 | const TEST: Fn = my_fn;
  |                  ^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `(dyn FnOnce() -> u8 + 'static)`
  = note: constant expressions must have a statically known size

error[E0277]: the size for values of type `(dyn FnOnce() -> u8 + 'static)` cannot be known at compilation time
 --> /tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs:4:14
  |
4 | const TEST2: (Fn, u8) = (TEST, 0);
  |              ^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `(dyn FnOnce() -> u8 + 'static)`
  = note: only the last element of a tuple may have a dynamically sized type

error[E0161]: cannot move a value of type `dyn FnOnce() -> u8`
 --> /tmp/icemaker_global_tempdir.NzAZRzGYMbbq/rustc_testrunner_tmpdir_reporting.SLbkYvmObPzd/mvce.rs:4:26
  |
4 | const TEST2: (Fn, u8) = (TEST, 0);
  |                          ^^^^ the size of `dyn FnOnce() -> u8` cannot be statically determined

thread 'rustc' panicked at compiler/rustc_const_eval/src/const_eval/eval_queries.rs:50:5:
assertion failed: layout.is_sized()
stack backtrace:
   0:     0x7f501e98caf6 - std::backtrace_rs::backtrace::libunwind::trace::hbec2edc838b433e2
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7f501e98caf6 - std::backtrace_rs::backtrace::trace_unsynchronized::h88c30257ee0dd166
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f501e98caf6 - std::sys_common::backtrace::_print_fmt::h9176654d509e3088
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f501e98caf6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h671feaee1776bf19
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f501e9dda7c - core::fmt::rt::Argument::fmt::h59605c83ae32d935
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/core/src/fmt/rt.rs:142:9
   5:     0x7f501e9dda7c - core::fmt::write::hf7af8dfcfd20db5b
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/core/src/fmt/mod.rs:1120:17
   6:     0x7f501e9814bf - std::io::Write::write_fmt::h92a5dbf489025e71
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/io/mod.rs:1846:15
   7:     0x7f501e98c8a4 - std::sys_common::backtrace::_print::h00080302d23998ae
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f501e98c8a4 - std::sys_common::backtrace::print::h6b23124445400245
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f501e98f5eb - std::panicking::default_hook::{{closure}}::hd67b2f9d36675a28
  10:     0x7f501e98f339 - std::panicking::default_hook::h10a39ef11c714e42
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/panicking.rs:292:9
  11:     0x7f502187a94c - std[5a8e7f94216a9138]::panicking::update_hook::<alloc[e14ab1f516d287ab]::boxed::Box<rustc_driver_impl[e279e4f1fedd9109]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f501e98fd50 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h37ba37dc48823c1d
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/alloc/src/boxed.rs:2030:9
  13:     0x7f501e98fd50 - std::panicking::rust_panic_with_hook::he3e3bc7c345bf54b
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/panicking.rs:786:13
  14:     0x7f501e98fa59 - std::panicking::begin_panic_handler::{{closure}}::h4bfbc62f377d59e6
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/panicking.rs:652:13
  15:     0x7f501e98cfd6 - std::sys_common::backtrace::__rust_end_short_backtrace::hc8487cf1bc04a693
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7f501e98f7c4 - rust_begin_unwind
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/panicking.rs:648:5
  17:     0x7f501e9d9f95 - core::panicking::panic_fmt::h88277cce839da186
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/core/src/panicking.rs:72:14
  18:     0x7f501e9da053 - core::panicking::panic::h1330285014ca2664
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/core/src/panicking.rs:144:5
  19:     0x7f5023a79bef - rustc_const_eval[b7e9a6566b8f5ff6]::const_eval::eval_queries::eval_body_using_ecx::{closure#0}
  20:     0x7f5023a76f23 - rustc_const_eval[b7e9a6566b8f5ff6]::const_eval::eval_queries::eval_in_interpreter
  21:     0x7f5022ffa938 - rustc_const_eval[b7e9a6566b8f5ff6]::const_eval::eval_queries::eval_to_allocation_raw_provider
  22:     0x7f5022ffa706 - rustc_query_impl[9d4c51777e2bb749]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4c51777e2bb749]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[446dd094ad85939e]::query::erase::Erased<[u8; 24usize]>>
  23:     0x7f502360ba51 - rustc_query_system[9b2ccfe41ae3bb1b]::query::plumbing::try_execute_query::<rustc_query_impl[9d4c51777e2bb749]::DynamicConfig<rustc_query_system[9b2ccfe41ae3bb1b]::query::caches::DefaultCache<rustc_middle[446dd094ad85939e]::ty::ParamEnvAnd<rustc_middle[446dd094ad85939e]::mir::interpret::GlobalId>, rustc_middle[446dd094ad85939e]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[9d4c51777e2bb749]::plumbing::QueryCtxt, false>
  24:     0x7f502360b62c - rustc_query_impl[9d4c51777e2bb749]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  25:     0x7f5022ffb60d - rustc_const_eval[b7e9a6566b8f5ff6]::const_eval::eval_queries::eval_to_const_value_raw_provider
  26:     0x7f5022ffb40c - rustc_query_impl[9d4c51777e2bb749]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4c51777e2bb749]::query_impl::eval_to_const_value_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[446dd094ad85939e]::query::erase::Erased<[u8; 24usize]>>
  27:     0x7f502360ba14 - rustc_query_system[9b2ccfe41ae3bb1b]::query::plumbing::try_execute_query::<rustc_query_impl[9d4c51777e2bb749]::DynamicConfig<rustc_query_system[9b2ccfe41ae3bb1b]::query::caches::DefaultCache<rustc_middle[446dd094ad85939e]::ty::ParamEnvAnd<rustc_middle[446dd094ad85939e]::mir::interpret::GlobalId>, rustc_middle[446dd094ad85939e]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[9d4c51777e2bb749]::plumbing::QueryCtxt, false>
  28:     0x7f502360b530 - rustc_query_impl[9d4c51777e2bb749]::query_impl::eval_to_const_value_raw::get_query_non_incr::__rust_end_short_backtrace
  29:     0x7f50235afe6b - <rustc_middle[446dd094ad85939e]::query::plumbing::TyCtxtEnsure>::const_eval_poly
  30:     0x7f50235adbd3 - rustc_hir_analysis[71d4294717f7f2b7]::check_crate
  31:     0x7f50236a2dd3 - rustc_interface[7ca501d7f8f3a34f]::passes::analysis
  32:     0x7f50236a2a29 - rustc_query_impl[9d4c51777e2bb749]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9d4c51777e2bb749]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[446dd094ad85939e]::query::erase::Erased<[u8; 1usize]>>
  33:     0x7f502393d4a5 - rustc_query_system[9b2ccfe41ae3bb1b]::query::plumbing::try_execute_query::<rustc_query_impl[9d4c51777e2bb749]::DynamicConfig<rustc_query_system[9b2ccfe41ae3bb1b]::query::caches::SingleCache<rustc_middle[446dd094ad85939e]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[9d4c51777e2bb749]::plumbing::QueryCtxt, false>
  34:     0x7f502393d209 - rustc_query_impl[9d4c51777e2bb749]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7f502394c719 - rustc_interface[7ca501d7f8f3a34f]::interface::run_compiler::<core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>, rustc_driver_impl[e279e4f1fedd9109]::run_compiler::{closure#0}>::{closure#0}
  36:     0x7f5023bbe705 - std[5a8e7f94216a9138]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[7ca501d7f8f3a34f]::util::run_in_thread_with_globals<rustc_interface[7ca501d7f8f3a34f]::util::run_in_thread_pool_with_globals<rustc_interface[7ca501d7f8f3a34f]::interface::run_compiler<core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>, rustc_driver_impl[e279e4f1fedd9109]::run_compiler::{closure#0}>::{closure#0}, core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>>::{closure#0}, core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>>
  37:     0x7f5023bbe532 - <<std[5a8e7f94216a9138]::thread::Builder>::spawn_unchecked_<rustc_interface[7ca501d7f8f3a34f]::util::run_in_thread_with_globals<rustc_interface[7ca501d7f8f3a34f]::util::run_in_thread_pool_with_globals<rustc_interface[7ca501d7f8f3a34f]::interface::run_compiler<core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>, rustc_driver_impl[e279e4f1fedd9109]::run_compiler::{closure#0}>::{closure#0}, core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>>::{closure#0}, core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[85b4fd477ce5bb21]::result::Result<(), rustc_span[adb133ac03280fc1]::ErrorGuaranteed>>::{closure#1} as core[85b4fd477ce5bb21]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7f501e999725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h1bff3efb61c3fa6b
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/alloc/src/boxed.rs:2016:9
  39:     0x7f501e999725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h83141cf549dced8a
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/alloc/src/boxed.rs:2016:9
  40:     0x7f501e999725 - std::sys::pal::unix::thread::Thread::new::thread_start::hbe9287fdd7c325bc
                               at /rustc/f8131a48a46ac3bc8a3d0fe0477055b132cffdc3/library/std/src/sys/pal/unix/thread.rs:108:17
  41:     0x7f501e7819eb - <unknown>
  42:     0x7f501e8057cc - <unknown>
  43:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.78.0-nightly (f8131a48a 2024-02-21) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [eval_to_allocation_raw] const-evaluating + checking `TEST2`
#1 [eval_to_const_value_raw] simplifying constant for the type system `TEST2`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 6 previous errors; 1 warning emitted

Some errors have detailed explanations: E0161, E0277, E0425, E0601.
For more information about an error, try `rustc --explain E0161`.

```

</p>
</details>

### PR DESCRIPTION
**Pull request by gurry**
Instead of returning the type of the last element we flip it around and return types of all elements _except_ the last one. This matches the the constraint on tuple types which says that only the last element can be unsized.

**Edit**
Returning all elements except the last was a mistake. We now return all the elements.

Fixes #121443

### CODE REVIEW

**Comment by gurry** [comment]
I'm not sure if this is the right fix. Please see my comment here: https://github.com/rust-lang/rust/issues/121443#issuecomment-1968418590 . Would appreciate any guidance.

**Comment by oli-obk** [comment]
`sized_conditions` and `sized_constraint_for_ty` are mostly duplicates of each other. we should probably deduplicate them $somehow.

They are meant to only look at the last type in structs and tuples (the "tail"), as that is the only thing that decides the sizedness of the entire type. This is sufficient in theory, because unsized types anywhere but the last field are unsupported and a hard error in wfcheck. Of course that doesn't help code that relies on wfcheck having bailed out already.

Maybe we should move such checks into stuff the `WellFormed` obligation checks, and then change the `wfcheck` check to just register that predcicate instead of manually checking it

**Comment by gurry** [comment]
Thanks. I'll implement according to your suggestions.

**Comment by oli-obk** [comment]
Those were just thoughts, I'm not sure this is a straightforward change. Tho the deduplication should probably be done on its own anyway

**Comment by gurry** [comment]
I have spent some time familiarizing myself with trait selection code and analyzed  #21443 again. 

What is happening in the issue is:

1. `check_item_type` in wfcheck adds both a `Sized` and a WF obligation for the tuple type in question: https://github.com/rust-lang/rust/blob/8b1af4c6610901b23be8d3311ed493cda0665983/compiler/rustc_hir_analysis/src/check/wfcheck.rs#L1252-L1260
2. During obligation processing the `Sized` obligation _does not_ fail and causes an `Ok` entry to be added to the global selection cache for the `Sized` trait predicate on the tuple type: https://github.com/rust-lang/rust/blob/f7cb53e54b0f2a403c12c123b5583a8d0d3ba3fe/compiler/rustc_trait_selection/src/traits/select/mod.rs#L396-L405 The WF obligation does result in `Sized` obligations registered for non-last fields of the tuple via it calls `wf::obligations`->`compute`: https://github.com/rust-lang/rust/blob/f7cb53e54b0f2a403c12c123b5583a8d0d3ba3fe/compiler/rustc_trait_selection/src/traits/wf.rs#L627-L633 which seems to be exactly what we want. However it has no impact whatsoever on the cache entry added above which stays put.
3. Later on, the fallback phase in typeck tries to select a `Sized` predicate on the tuple type and finds the one added to the cache above. As a result typeck passes for the type and the ICE occurs during const eval.

Based on the above analysis, the fix in `sized_conditions` does seem to be the best solution as it should prevent the bad entry from going into the cache. However, in my earlier commit I messed up by returning _only the non-last_ fields from `size_conditions` whereas I should've returned all the fields.

Now I have pushed a new commit in which `sized_conditions` which does just that. It causes the `Sized` obligation on the tuple fail during confirmation: https://github.com/rust-lang/rust/blob/f7cb53e54b0f2a403c12c123b5583a8d0d3ba3fe/compiler/rustc_trait_selection/src/traits/select/mod.rs#L1270-L1279 which adds an `Err` entry to the selection cache. 

#21443 is now fixed, the ICEs in codegen we saw earlier are gone and all the tests are green. @oli-obk Please let me know what you think and whether this approach can be viable at all.

File: tests/ui/impl-trait/dyn-trait-return-should-be-impl-trait.rs [hunk 1]
```diff
@@ -5,11 +5,11 @@ impl Trait for Struct {}
...
 fn fuz() -> (usize, Trait) { (42, Struct) }
-//~^ ERROR E0277
-//~| ERROR E0308
+//~^ ERROR E0308
+//~| ERROR E0277
 fn bar() -> (usize, dyn Trait) { (42, Struct) }
-//~^ ERROR E0277
-//~| ERROR E0308
+//~^ ERROR E0308
+//~| ERROR E0277
```
Comment (oli-obk) [inline tests/ui/impl-trait/dyn-trait-return-should-be-impl-trait.rs:12]: now-unnecessary diff

**Review by oli-obk — COMMENTED** [review]
I think this is correct. I am worried there is something this will break out in the wild, but I think it's different from how we accept `fn() -> str` function pointers in the type system, even if they can't be constructed or used.

@rust-lang/types can you think of any issues with checking that in order for a tuple to be sized, all its fields need to be sized (instead of just checking the last field)

**Comment by lcnr** [comment]
Checking that all tuple fields are `Sized` for `(T, U): Sized` is not wrong, and would even match the user written impl (if it were possible to write `Sized` impls), however, it is likely less efficient and may not fix the underlying issue in #121443. 

From what I can tell the root issue here is that we do not abort evaluation even though typeck failed. I don't know why this is the case but both `TEST` and `TEST2` should fail typeck as `(T, U): Sized` already requires `T: Sized` for wf, as explained by @gurry.

**Comment by oli-obk** [comment]
> From what I can tell the root issue here is that we do not abort evaluation even though typeck failed.

it's not a typeck error, but a wfcheck error, and that happens in a separate query invocation tree. There is no way to carry information from wfcheck to typeck without causing lots of cycles.

**Comment by lcnr** [comment]
why does typeck not check that the return type is well-formed?

**Comment by gurry** [comment]
> why does typeck not check that the return type is well-formed? We should check that in typeck as well 🤔

Not well formed, but typeck does check that the the return type is sized. It does so by adding a `Sized` obligation over here: https://github.com/rust-lang/rust/blob/7606c13961ddc1174b70638e934df0439b7dc515/compiler/rustc_hir_typeck/src/lib.rs#L256 which is then checked here: https://github.com/rust-lang/rust/blob/7606c13961ddc1174b70638e934df0439b7dc515/compiler/rustc_hir_typeck/src/lib.rs#L266

Unfortunately this check passes because it finds the `Sized` predicate entry setting in selection cache which was added earlier during wfcheck (see my previous comment).

**Comment by oli-obk** [comment]
> Maybe we should move such checks into stuff the `WellFormed` obligation checks, and then change the `wfcheck` check to just register that predcicate instead of manually checking it

We could still try this scheme (and then make typeck do the wfchecks instead of wfcheck)

**Comment by gurry** [comment]
> Maybe we should move such checks into stuff the `WellFormed` obligation checks, and then change the `wfcheck` check to just register that predcicate instead of manually checking it

Sorry, I didn't fully understand what you meant😟(mostly due to my new-ness to this area).

wfcheck doesn't seem to be doing any manual checks for tuples even now. All it does is add `WellFormed` obligations. When the WF obligation is processed in `process_obligations` it  doesn't do any direct WF checks either. It simply adds child obligations for the sizedness of all non-last fields over here: https://github.com/rust-lang/rust/blob/f7cb53e54b0f2a403c12c123b5583a8d0d3ba3fe/compiler/rustc_trait_selection/src/traits/wf.rs#L627-L633 which is exactly what we would like. However, these child obligations do not have the desired effect. They do not prevent a `Sized` trait predicate for the tuple type from being added to the cache.

Please elaborate a bit more on what you had in mind and I'll investigate further.

File: tests/ui/impl-trait/dyn-trait-return-should-be-impl-trait.rs (hunk 1 above)
Comment (gurry) [reply tests/ui/impl-trait/dyn-trait-return-should-be-impl-trait.rs:12]: Ooh

**Comment by oli-obk** [comment]
> They do not prevent a `Sized` trait predicate for the tuple type from being added to the cache.

that's fine. What lcnr was proposing was to liberally add additional `WellFormed` obligations in typeck, instead of just adding `Sized` obligations.

File: tests/ui/impl-trait/dyn-trait-return-should-be-impl-trait.rs (hunk 1 above)
Comment (gurry) [reply tests/ui/impl-trait/dyn-trait-return-should-be-impl-trait.rs]: Fixed

**Comment by gurry** [comment]
> > They do not prevent a `Sized` trait predicate for the tuple type from being added to the cache.
> 
> that's fine. What lcnr was proposing was to liberally add additional `WellFormed` obligations in typeck, instead of just adding `Sized` obligations.

If we only register a `WellFormed` obligation in typeck, I'm afraid it won't have any effect, because the previous `WellFormed` obligation registered by wfcheck didn't either.  The main problem here is the bad `Sized` entry sitting in the selection cache (which by way comes from from the _`Sized`_ obligation registered by wfcheck). If we want to do any additional checks in typeck, we will have to somehow get rid of that cache entry first.

**Comment by oli-obk** [comment]
> The main problem here is the bad `Sized` entry sitting in the selection cache (which by way comes from from the _`Sized`_ obligation registered by wfcheck). If we want to do any additional checks in typeck, we will have to somehow get rid of that cache entry first.

I assumed that `WellFormed((T, U))` got expanded to various things, among them `T: Sized`, which would be exactly what we'd need

**Comment by oli-obk** [comment]
> because the previous `WellFormed` obligation registered by wfcheck didn't either

how did we end up getting an error then? We do get the error `the size for values of type `(dyn FnOnce() -> u8 + 'static)` cannot be known at compilation time` before we ICE

**Comment by oli-obk** [comment]
if a type is not well-formed, we can probably prove all kinds of broken things for it. So the `(T, U): Sized` check *assumes* that `(T, U)`  is well formed, and thus it only needs to check `U` for sizedness. Any field but the last does not affect the `Sized`ness of the tuple, it affects the well-formedness.

**Comment by oli-obk** [comment]
That said, ideally we'd just do this properly with the change that you did here, but that is only part of the story, there may be other ways in which those tuples (or other types) may not be well formed, and cause ICEs subsequently.

That said, let's at least check the claim that this change is a perf issue

@bors try @rust-timer queue

**Comment by gurry** [comment]
> I assumed that `WellFormed((T, U))` got expanded to various things, among them `T: Sized`, which would be exactly what we'd need

Yes it did. The processing of `WellFormed((T, U))`  in `process_obligation` constitutes of calling `wf::obligations` which in turn calls `compute` which does this:https://github.com/rust-lang/rust/blob/f7cb53e54b0f2a403c12c123b5583a8d0d3ba3fe/compiler/rustc_trait_selection/src/traits/wf.rs#L627-L633 i.e. it adds `Sized` obligations for all non-last fields which would be `T: Sized` in our case.

And this `T: Sized` does fail. Unfortunately it doesn't prevent an `Ok` entry for `(T, U): Sized` from being added to the selection cache.

Ideally the failure of a child obligations should add an `Err` for the parent obligation. But the code in `process_obligations` doesn't do any such thing.

The only place where something like this occurs is in `confirm_candidate` deeper in the call hierachy. But at that place the above  parent-child relation is not known. The only parent child relation known there is the one returned by `sized_conditions`. Hence this change.

**Comment by oli-obk** [comment]
We don't actually need `(T, U): Sized` to fail, any failure, as long as it happens in typeck, will taint the typeck results and prevent const eval from running at all.

Are you saying that the implied `T: Sized` obligation failing will not cause an error if we're inserting a `WellFormed((T, U))` obligation?

**Comment by gurry** [comment]
> We don't actually need `(T, U): Sized` to fail, any failure, as long as it happens in typeck, will taint the typeck results and prevent const eval from running at all.

True. The immediate need is only to have typeck fail. We don't really care about `(T, U): Sized` directly. But what makes `(T, U): Sized` of interest is that _if_ it fails it guarantees that typeck will also fail (see below). 

> Are you saying that the implied `T: Sized` obligation failing will not cause an error if we're inserting a `WellFormed((T, U))` obligation?

`T: Sized` is indeed failing and it is showing an error to the user, which is the one you mentioned above. It is just not causing typeck to fail. That is because typeck is NOT checking for `T: Sized` failure. It is checking for `(T, U): Sized` failure.

And all of this is happening regardless of the presence of WellFormed((T, U))`  i.e. it does not have any impact. So we can ignore it

But this given me an idea - what if we made typeck to check for `T: Sized` instead of just (T, U): Sized`? In other words, what if typeck checked all the non-last fields of a tuple for sizedness?

Now that it is slowly dawning on me, isn't that what you were suggesting originally?

**Comment by gurry** [comment]
> if a type is not well-formed, we can probably prove all kinds of broken things for it. So the `(T, U): Sized` check _assumes_ that `(T, U)` is well formed, and thus it only needs to check `U` for sizedness. Any field but the last does not affect the `Sized`ness of the tuple, it affects the well-formedness.

That means `(T, U): Sized` implies `(T, U)` is WF implies `T: Sized`

Or, transitively, `(T, U): Sized`  implies  `T: Sized`

If so, `T: Sized` NOT holding must imply `(T, U): Sized` should NOT hold. This is exactly what we're trying to ensure with this PR.

**Comment by lcnr** [comment]
We imply that `(T, U): Sized` is *correct* if `WellFormed((T, U))` holds. This means that `(T, U): Sized` is unspecified if `WellFormed((T, U))` does not hold.

`(T, U): Sized` does not imply `WellFormed((T, U))`.

**Comment by oli-obk** [comment]
> what kind of cycles do you get if you also taint the MIR if wfcheck failed?

Mostly const eval cycles iirc, but let's actually try it and document it xD

**Comment by gurry** [comment]
> We imply that `(T, U): Sized` is _correct_ if `WellFormed((T, U))` holds. This means that `(T, U): Sized` is unspecified if `WellFormed((T, U))` does not hold. `(T, U): Sized` does not imply `WellFormed((T, U))`.

Just curious. What does `(T, U): Sized` being _correct_ mean and how is it different from it _holding_?

**Comment by oli-obk** [comment]
> Just curious. What does `(T, U): Sized` being _correct_ mean and how is it different from it _holding_? Aren't they the same thing?

The situation right now is that `(T, U): Sized` holds, even if `T` is unsized, but it is not correct, because `(T, U)` is not well-formed if `T` is unsized.

**Comment by gurry** [comment]
@oli-obk @lcnr I have implemented the alternate solution proposed by you which relies on doing WF check in typeck: #122078. If that solution is better we can abandon this PR.

**Comment by oli-obk** [comment]
Jup, I prefer the other PR. Thanks!
